pkg/lib/cockpit-components-file-autocomplete.jsx

```diff
@@ -180,6 +180,12 @@ class FileAutoComplete extends React.Component {
         });
     }
 
+    removeDuplicatePaths(displayFiles) {
+        return displayFiles.filter((obj, pos, arr) => {
+            return arr.map(dict => dict.path).indexOf(obj.path) === pos;
+        });
+    }
+
```

Comment:
This is a quadratic algorithm that runs in every render(), so quite expensive. IMHO it's better to fix the state to not contain duplicate keys in the first place. But how can that happen? Does `updateFiles()` really produce duplicate files in one directory? If so, then there's a bug there, and let's rather fix it rather than papering over it. Do you haven an example how such a duplicated file name looks like?

Reply:
Yes, here is the dump of the ```this.state.displayFiles``` variable from which we use that path values for keys here

https://github.com/cockpit-project/cockpit/blob/eef5ce0ef699977d22e0f8048615b93aa3c2153a/pkg/lib/cockpit-components-file-autocomplete.jsx#L275

``` python
[
    {'event': 'present', 'path': 'boot/', 'type': 'directory'},
    {'event': 'present', 'path': 'dnsmasq/', 'type': 'directory'},
    {'event': 'present', 'path': 'filesystems/', 'type': 'directory'},
    {'event': 'present', 'path': 'images/', 'type': 'directory'},
    {'event': 'present', 'path': 'network/', 'type': 'directory'},
    {'event': 'present', 'path': 'novell.iso', 'type': 'file'},
    {'event': 'present', 'path': 'qemu/', 'type': 'directory'},
    {'event': 'present', 'path': 'swtpm/', 'type': 'directory'},
    {'event': 'attribute-changed', 'path': '/var/lib/libvirt/novell.iso'
     , 'tag': '1:21062602-1542365534.25178631'},
    {'event': 'attribute-changed', 'path': '/var/lib/libvirt/novell.iso'
     , 'tag': '1:21062602-1542365534.25178631'}
]
```

You can see that the ```/var/lib/libvirt/novell.iso``` path exists twice.

Reply:
@martinpitt does this tell you something? Also I don't get why the last two have ```attribute-changed``` event and the others just ```present```.

Reply:
Correct. The fix is trivial in `channel.addEventListener("message",`. But I'll make a separate PR with a specific integration test and also update doc/protocol.md. That only documents `present`, but not other fields.